### DISCUSSION TITLE
What’s the list of commands inside a sieve script that we can use with the storage and directories backends?

### DISCUSSION THREAD

**Discussion by 11notes** [stalwartlabs/smtp-server#49] [Q&A]
Is there a list of documented commands we can use inside sieve for the different types for storage and directories? Like ```key_exists``` for lookups for example but more? Can we store information too and not just query? Like write a key to Redis in a Sieve?

**Comment by mdecimus** [comment]
Hi,

Stalwart uses expressions in two places; Sieve scripts (with the `eval` command) and from dynamic rules in the config file. Although their syntax is exactly the same, Sieve expressions support more functions than the config expressions. Currently only the config expression functions are [documented](https://stalw.art/docs/configuration/expressions/functions). All the functions you need to interact with databases and directories are documented on that page. It is on the roadmap to improve the documentation for developers (not only Sieve expression functions but also the REST API) but in the meantime you can take a look at [this file](https://github.com/stalwartlabs/mail-server/blob/4f375a48830943fed58fe3b6e90f1ad4849253a7/crates/common/src/scripts/plugins/mod.rs#L54) and [this file](https://github.com/stalwartlabs/mail-server/blob/4f375a48830943fed58fe3b6e90f1ad4849253a7/crates/common/src/scripts/functions/mod.rs#L40) to see all the additional commands supported by Sieve expressions.
